File: 0912-sort-an-array/0912-sort-an-array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortArray = function(nums) {
    // Base case: if the array has 0 or 1 element, it's already sorted
    if (nums.length <= 1) {
        return nums;
    }

    // Split the array into two halves
    const mid = Math.floor(nums.length / 2);
    const left = sortArray(nums.slice(0, mid));
    const right = sortArray(nums.slice(mid));

    // Merge the sorted halves
    return merge(left, right);
};

// Helper function to merge two sorted arrays
function merge(left, right) {
    const sorted = [];
    let i = 0, j = 0;

    // Compare elements from left and right arrays and merge them in sorted order
    while (i < left.length && j < right.length) {
        if (left[i] <= right[j]) {
            sorted.push(left[i]);
            i++;
        } else {
            sorted.push(right[j]);
            j++;
        }
    }

    // If there are remaining elements in left, add them
    while (i < left.length) {
        sorted.push(left[i]);
        i++;
    }

    // If there are remaining elements in right, add them
    while (j < right.length) {
        sorted.push(right[j]);
        j++;
    }

    return sorted;
}

File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        if len(nums) <= 1:
            return nums
        
        mid = len(nums) // 2
        left = self.sortArray(nums[:mid])
        right = self.sortArray(nums[mid:])
        
        return self.merge(left, right)
    
    def merge(self, left: List[int], right: List[int]) -> List[int]:
        sorted_list = []
        i = j = 0
        
        while i < len(left) and j < len(right):
            if left[i] <= right[j]:
                sorted_list.append(left[i])
                i += 1
            else:
                sorted_list.append(right[j])
                j += 1
        
        sorted_list += left[i:]
        sorted_list += right[j:]
        
        return sorted_list
